core/src/test/java/org/apache/calcite/test/RelOptRulesTest.java

```diff
@@ -6827,6 +6828,62 @@ protected MyProjectRule(Config config) {
         .check();
   }
 
+  /**
+   * Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-4317">[CALCITE-4317]
+   * RelFieldTrimmer after trimming all the fields in an aggregate
+   * should not return a zero field Aggregate</a>. */
+  @Test void testProjectJoinTransposeRuleOnAggWithNoFieldsWithTrimmer() {
```

Comment:
Shouldn't this new test be added inside RelFieldTrimmerTest rather than RelOptRulesTest?

Reply:
There were 2 parts to be tested here:
1. Trimmer handles aggregate node if all the fields have been trimmed.
2. Project-Join transpose rule handles the trimmed node properly.

So i had to add the test in RelOptRulesTest for 2, for which i had to trim the node first, so it covers 1 as well. If you think i should add a separate test for 1 in RelFieldTrimmerTest, that can be done.

Reply:
Ok, thanks for the explanation. IMO there is no need to duplicate the test, we can leave it just here.

Reply:
Hi, @rubenada I mean that can be test evidence for this patch and this block CALCITE-4381, I hope it can merge as soon as possible.

Reply:
Hi @a-rafay, can you represent this ut by SQL, it may be more clearer

Reply:
@Aaaaaaron sorry for the confusion, my previos comment was actually replying  @a-rafay 's last comment
https://github.com/apache/calcite/pull/2211#discussion_r518333162

Reply:
@Aaaaaaron my test case was slightly different but it seems like similar to CALCITE-4381, i couldn't reproduce the query we had since we do a lot of transformation before getting to that state so i started off with creating a RelNode and applying the rule on that. If this patch fixes CALCITE-4381, then it shouldn't matter what test case we used if the nature of test case is the same.
You also asked about the progress of this PR. I am just waiting for this to get reviewed so we can merge this.

Reply:
@a-rafay "then it shouldn't matter what test case we used if the nature of test case is the same." yes, this is what I mean, @rubenada can you help review and merge this PR, I've been met this times, thanks!

Reply:
Would it be possible to have the test defined in terms of SQL:
```
SELECT name FROM (SELECT count(*) cnt_star, count(empno) cnt_en FROM sales.emp)
cross join sales.dept limit 10
```
rather than RelBuilder?

Reply:
> Would it be possible to have the test defined in terms of SQL:
> 
> ```
> SELECT name FROM (SELECT count(*) cnt_star, count(empno) cnt_en FROM sales.emp)
> cross join sales.dept limit 10
> ```
> 
> rather than RelBuilder?

I can add a test using the query, but that will only test if we are not getting emtpy Agg after trimming. The second issue was to test if the ProjectJoinTransposeRule will not throw an exception, which we will not be able to test since the resulting RelNode does not match the rule. In our environment, we get the transformed node after applying some rules/transformation which i was unable to come up with straight from the query. Hence i used RelBuilder to start off with the node that will satisfy both parts of the ticket. Non-empty agg node after trimming, and firing ProjectJoinTransposeRule.

Reply:
Ok, thanks for the explanation, I guess the RelBuilder test suffices.